### DISCUSSION TITLE
Toast Timeout Global Configuration

### DISCUSSION THREAD

**Discussion by to-marss** [Blazored/Toast#201] [Q&A] [answered]
Hi,

I am trying to use Blazor Toast for .NET 6.0 Blazor Server. I would like to know if there is a way to configure timeout values globally for each type. For e.g. Info = 5 sec, Error = 10 sec.

Current version allows to configure Timeout once globally at MainLayout.razor page and can be overridden locally when the toastservice.ShowInfo() or whichever is called. Let's say there are 1000 places toaster is used along with timeout for 5 seconds for INFO and 10 seconds for ERROR. If this needs a change to have ERROR toast timeout as 15 seconds, it requires a change across app. In order to prevent this, it would be good to have a global config timeout per type. This way a change would be in one place and not across.

Can this be done or extended with current library? Please advise.

NOTE: Just a heads up, getting this global config can be from .json file. This would allow app users to control the toaster timeout config based on types hence the need arise. I would just need the help how we can make use of this library to achieve this :)

**Comment by Cvijo** [comment, marked as answer]
I think a workaround would be to create your own service MyToastService, inject Blazored Toast Service and create your own ShowError, ShowWarrning, ShowInfo .. etc and put on ShowError 15 sec timeout, and pass the rest of the parameters to blazored  toast service.

**Reply by to-marss** [reply]
@Cvijo  thanks for prompt response. I was thinking of the same workaround because we can't extend ToastService and override functions. As a suggested workaround I need to inject IToastService and create CustomToastService entirely. So. lookin for few alternatives. Like, can we listen to these events showInfo, showerror etc at one place globally when it's called and make changes to TestInstanceSettings Timeout?  i.e. time set based on my json config when show event is fired. Please advise. 
@chrissainty , @peterblazejewicz , @thoemmi . @MarkStega ++

**Reply by chrissainty** [reply]
@to-marss Your only option right now is @Cvijo suggestion. The current settings don't allow timeouts to be set per toast type.

**Reply by to-marss** [reply]
thanks! @Cvijo @chrissainty 
Also, is there a way to disable timeout for some toasts? Setting timeout to 0 doesn't work. For e.g. I need to show some toast messages permanently, forcing users to close on their own and other toast messages can have default timeout. How to do this?

**Reply by MarkStega** [reply]
@to-marss @chrissainty @simonziegler

We used a fork of this project to form the basis for the toast implemented as MBToast in [Material.Blazor](https://github.com/Material-Blazor/Material.Blazor). While we don't have a timeout per level, we do have a ShowToast() method that has the ability to set a timeout for each invocation. I use that in several projects to show really short toasts for things like logon success and ones that do not timeout and require use of the dismiss button for fatal errors.

Feel free to copy the code (or use in Material.Blazor). M.B 3.x are stable releases that are .Net 6 compatible. M.B 4.x is in preview and is .Net 7 compatible.

**Reply by to-marss** [reply]
@MarkStega thanks for quick response. Sounds like a good idea but I can't make a change to migrate to [Material.Blazor] (https://github.com/Material-Blazor/Material.Blazor) at this point of time. Any workaround possible with current Blazored.Toast library for this please?

**Reply by MarkStega** [reply]
@to-marss

You could extract the code from Material.Blazor that is toast code and just incorporate it into your project. Alternatively, you could take a fork of Blazored.Toast and do the same (and even contribute back to the .project)

**Comment by to-marss** [comment]
@chrissainty @simonziegler @Cvijo while I wait for the workaround above. I am facing an issue with timeout. For e.g. ShowInfo with settings.Timeout = 30 doesn't wait for 30 seconds. It timeout after 5 seconds which I believe is the default value. This is a show stopper concurrent to this task. 

Should I file an issue for this?

private void OnShowInfo()
{
            string message = "Info message from home.razor.cs";
            toastService.ShowInfo(message, settings => { settings.Timeout = 30; settings.ShowProgressBar = false; });
}

![image](https://user-images.githubusercontent.com/117723216/218765870-a7baa70a-8c94-4b3b-9aaf-625a04e87e61.png)

**Reply by chrissainty** [reply]
What workaround are you waiting on? I just want to make sure I've not missed anything. 

The bug you mentioned has been identified and is tracked via https://github.com/Blazored/Toast/issues/195

**Reply by to-marss** [reply]
@chrissainty 
1. Regarding issue #195 , any ETA please? When can I expect the releas, current development is on hold due to this :(
2. Workaround to disable timeout for some toasts? Setting timeout to 0 doesn't work. For e.g. I need to show some toast messages permanently, forcing users to close on their own and other toast messages can have default timeout ?

**Reply by chrissainty** [reply]
@to-marss Are you planning on sponsoring the project?

**Reply by to-marss** [reply]
Nice library you built so I would love to. I can wait on the disable timeout workaround but to get this issue fixed and available on live soon would help much. If not possible, unfortunately I will have to find other library to get this going which I don't want to :)

**Reply by Cvijo** [reply]
@to-marss, I am just loving Blazored and the effort and nuggets they provide and I am not a contributor to this nugget but ill try to explain why this is not just simple to add.

If you ask me about adding timeout = 0 to stay infinitive would require some design changes. 

- First, there must be ShowCloseButton true, if you don't have a close button how will you close toast, you can set it to false in options, so if you are setting it to false, they should override your settings and show it anyway and people will tell its a bug, i put in options not to show close button.
- Second, you can style and add your own custom component inside toast, and how will you implement the close method? In this case, you are setting your own component and should use OnClick Action event to close the toast, but you don't have an instance of this toast and you can only call to close all instances of certain Levels, not the specific one. And then you are losing the option to have OnClick do something else (open modal, navigate to new page, etc)
- Third, you can add your own CSS styles to existing toast component that can hide/overlap close button

etc... etc 

There are many use cases that will not work or will produce a lot of other problems if they just add if timeout = 0, leave toast displayed :), but yes it could be done by doing some redesign

and there is no workaround for this atm, you can only add your own MyToastService as i mention to you in the first post , and add your own ShowError method and pass settings to be timeout = Int32.MaxValue, and set ShowCloseButton = true, something like: 

`toastService.ShowError("your error message", settings => { settings.Timeout = Int32.MaxValue; settings.ShowProgressBar = false; settings.ShowCloseButton = true ; });`

**Reply by to-marss** [reply]
@Cvijo absolutely yes! thanks for the clarification and the suggested workaround will do, thanks again :)

Just waiting for #192 release and this will make my day 👍  thanks everyone for the timely help..

**Reply by to-marss** [reply]
@chrissainty  @Cvijo as #192 is held up for release. Do we have any workaround for it? I don't see any other option and doesn't want to end up reusing the entire library in my code. Please provide ur best help here!

**Reply by Cvijo** [reply]
@to-marss  i am sure @chrissainty  will create a new release (update for nuget) when everything is merged and ready.

If you are developing now just continue to use it like it is working (your toasts will be 10 sec all or whatever you set up globally) and when the new NuGet is out everything will work as expected.

If you are insisting on a workaround where you are setting your <BlazoredToasts ...options /> probably on MainLayout use the Timeout option as binding to the variable that you will change. (to be honest, I don't see the point)
